### PR TITLE
add draft exit code to kubectl-conventions

### PR DESCRIPTION
**Pull request by shiywang**
@fabianofranz @smarterclayton @pwittrock @kubernetes/sig-cli-maintainers

### CODE REVIEW

File: contributors/devel/kubectl-conventions.md [hunk 1]
```diff
@@ -375,6 +376,21 @@ this is not a mandatory structure and not every command is implemented this way,
 but this is a nice convention so try to be compliant with it. As an example,
 have a look at how [kubectl logs](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/logs.go) is implemented.
 
+## Exit code conventions
+
+Generally, for all the commands exit code, result of `zero` means succeed and `non-zero` means error.
+
+For some declarative commands like `apply` may have a situation which command exited success but without actual changes, we provide a flag `--error-unchanged` to specify the exit code to 2 if necessary.
```
Comment (liggitt) [inline contributors/devel/kubectl-conventions.md]: For idempotent operations, requiring no changes *is* success (apply is a "make it so" command… if it is already so, a non-zero exit code makes sense)

File: contributors/devel/kubectl-conventions.md [hunk 2]
```diff
@@ -266,7 +267,7 @@ rather than "RESOURCE" or "KIND"
 
 * Use "NAME" for resource names
 
-## kubectl Factory conventions
+## Kubectl Factory conventions
```
Comment (smarterclayton) [inline contributors/devel/kubectl-conventions.md]: This is unnecessary, it's acceptable to use `kubectl` lower case at the beginning of a title when referring to a CLI command.

File: contributors/devel/kubectl-conventions.md (hunk 1 above)
Comment (smarterclayton) [reply contributors/devel/kubectl-conventions.md]: Yeah, please rephrase as Jordan mentioned.  Idempotent ("make-it-so") commands should return 0 when successful (even if no changes were provided).  For commands that are not explicitly expected to be idempotent, consumers can request treating "already-so" as "make-it-so" via flag.

File: contributors/devel/kubectl-conventions.md [hunk 3]
```diff
@@ -375,6 +376,21 @@ this is not a mandatory structure and not every command is implemented this way,
...
+For some declarative commands like `apply` may have a situation which command exited success but without actual changes, we provide a flag `--error-unchanged` to specify the exit code to 2 if necessary.
+
+For other commands like `create` that return error when unchanged, we have the flag `--ignore-unchanged` to make it return 0 and no errors.
+
+
+| Exit Code Number | Meaning | Enable |
+| :---             | :---    | :---    |
+| 0                | Command exited success  | By default, By flag `--ignore-unchanged` |
+| 1                | Command exited for general errors  | By default |
+| 2                | Command exited success but without changes | By flag `--error-unchanged`|
```
Comment (smarterclayton) [inline contributors/devel/kubectl-conventions.md]: Probably rephrase this to `Command was successful, but the user requested a distinct exit code when no change was made`.

Also, we shouldn't be using 2 which is reserved as a bash shell error.  We should use 3 or higher.  See http://www.tldp.org/LDP/abs/html/exitcodes.html

File: contributors/devel/kubectl-conventions.md (hunk 1 above)
Comment (shiywang) [reply contributors/devel/kubectl-conventions.md]: combine two comments above then add a little redundancy and parallelism, ptal

**Comment by shiywang** [comment]
combine two comments above then add a little redundancy and parallelism, @liggitt @smarterclayton @fabianofranz ptal

**Comment by castrojo** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/community/1165)

**Comment by mengqiy** [comment]
@shiywang Have you implement this?

**Comment by shiywang** [comment]
@mengqiy here is my pr: https://github.com/kubernetes/kubernetes/pull/52647
